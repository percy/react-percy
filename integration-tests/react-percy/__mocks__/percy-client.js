export const createBuild = jest.fn(({ resources }) =>
  Promise.resolve({
    body: {
      data: {
        id: 'test-build',
        attributes: {
          'web-url': 'http://percy.local/test-build',
        },
        relationships: {
          'missing-resources': {
            data: resources,
          },
        },
      },
    },
  }),
);

export const createSnapshot = jest.fn(() =>
  Promise.resolve({
    body: {
      data: {
        id: 'test-snapshot',
      },
    },
  }),
);

export const finalizeBuild = jest.fn(() => Promise.resolve());

export const finalizeSnapshot = jest.fn(() => Promise.resolve());

export const makeResource = jest.fn(resource => {
  const sha = `sha-${resource.resourceUrl}-sha`;
  return Object.assign({}, resource, {
    id: sha,
    sha,
  });
});

export const uploadResources = jest.fn(() => Promise.resolve());

export default class FakePercyClient {
  createBuild = createBuild;
  createSnapshot = createSnapshot;
  finalizeBuild = finalizeBuild;
  finalizeSnapshot = finalizeSnapshot;
  makeResource = makeResource;
  uploadResources = uploadResources;
}
