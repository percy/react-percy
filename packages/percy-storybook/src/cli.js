import getStories from './getStories';
import getStaticAssets from './getStaticAssets';
import selectStories from './selectStories';
import uploadStories from './uploadStories';

import ApiClient from 'react-percy-api-client';
import createDebug from 'debug';

const debug = createDebug('percy-storybook');

// eslint-disable-next-line import/prefer-default-export
export async function run() {
    const stories = getStories();
    debug('stories %o', stories);

    const selectedStories = selectStories(stories);
    debug('selectedStories %o', selectedStories);

    const { storyHtml, assets } = getStaticAssets();
    debug('assets %o', assets);

    const client = new ApiClient(
      process.env.PERCY_TOKEN,
      process.env.PERCY_API
    );

    return uploadStories(client, selectedStories, storyHtml, assets);
}
