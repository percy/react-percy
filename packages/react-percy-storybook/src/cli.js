import * as args from './args';

import getStories from './getStories';
import getStaticAssets from './getStaticAssets';
import getWidths from './getWidths';
import selectStories from './selectStories';
import uploadStorybook from './uploadStorybook';

import ApiClient from '@percy-io/react-percy-api-client';
import createDebug from 'debug';

import yargs from 'yargs';

const debug = createDebug('percy-storybook');
const VERSION = require('../package.json').version;


// eslint-disable-next-line import/prefer-default-export
export async function run(argv) {
    argv = yargs(argv)
        .usage(args.usage)
        .help()
        .alias('help', 'h')
        .options(args.options)
        .epilogue(args.docs)
        .default('build_dir', 'storybook-static')
        .argv;

    if (argv.help) {
        yargs.showHelp();
        process.on('exit', () => process.exit(1));
        return;
    }

    if (argv.version) {
        process.stdout.write(`v${VERSION}\n`);
        process.on('exit', () => process.exit(0));
        return;
    }

    const widths = getWidths(argv.widths);

    const options = {
        debug: argv.debug,
        buildDir: argv.build_dir
    };

    if (!process.env.PERCY_TOKEN) {
        // eslint-disable-next-line no-console
        console.log('The PERCY_TOKEN environment variable is missing. Exiting.');
        return;
    }

    if (!process.env.PERCY_PROJECT) {
        // eslint-disable-next-line no-console
        console.log('The PERCY_PROJECT environment variable is missing. Exiting.');
        return;
    }

    const { storyHtml, assets, storybookJavascriptPath } = getStaticAssets(options);
    // debug('assets %o', assets);

    getStories(assets[storybookJavascriptPath], options).then((stories) => {
        debug('stories %o', stories);

        const selectedStories = selectStories(stories);
        debug('selectedStories %o', selectedStories);

        if (selectedStories.length === 0) {
            console.log('No stories were found.'); // eslint-disable-line no-console
            return;
        }

        const client = new ApiClient(
          process.env.PERCY_TOKEN,
          process.env.PERCY_API
        );

        return uploadStorybook(client, selectedStories, widths, storyHtml, assets);
    }).catch((reason) => {
        console.log('Error encountered taking snapshots.'); // eslint-disable-line no-console
        console.log(reason); // eslint-disable-line no-console
    });
}
