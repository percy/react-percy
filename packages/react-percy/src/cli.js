import * as args from './args';
import chalk from 'chalk';
import readPercyConfig from '@percy/react-percy-config';
import runPercy from '@percy/react-percy-ci';
import yargs from 'yargs';

process.env.NODE_ENV = process.env.NODE_ENV || 'development';

const VERSION = require('../package.json').version;

function formatError(error) {
  if (error.error && error.error.errors && error.error.errors.length) {
    return error.error.errors.map(e => e.detail).join(', ');
  }
  return error.stack || error;
}

// eslint-disable-next-line import/prefer-default-export
export function run(argv, rootDir) {
  argv = yargs(argv)
    .usage(args.usage)
    .help()
    .alias('help', 'h')
    .options(args.options)
    .epilogue(args.docs).argv;

  if (argv.help) {
    yargs.showHelp();
    process.on('exit', () => process.exit(1));
    return;
  }

  if (argv.version) {
    process.stdout.write(`v${VERSION}\n`);
    process.on('exit', () => process.exit(0));
    return;
  }

  if (process.env.PERCY_ENABLE === '0') {
    process.stdout.write('Percy is disabled by PERCY_ENABLE=0 environment variable. Skipping.\n');
    process.on('exit', () => process.exit(0));
    return;
  }

  if (!argv.debug && !process.env.PERCY_TOKEN) {
    process.stdout.write(
      chalk.bold.red('PERCY_TOKEN') + chalk.red(' environment variable must be set.\n'),
    );
    process.on('exit', () => process.exit(1));
    return;
  }

  const packageRoot = rootDir || process.cwd();

  const percyConfig = readPercyConfig(packageRoot, argv.debug);

  return runPercy(percyConfig, process.env.PERCY_TOKEN)
    .then(() => {
      process.on('exit', () => process.exit(0));
    })
    .catch(err => {
      if (err.name === 'StatusCodeError') {
        if (err.statusCode >= 500) {
          // eslint-disable-next-line no-console
          console.log(chalk.bold.yellow('Percy Error: %s'), formatError(err));
          process.on('exit', () => process.exit(0));
        } else {
          // eslint-disable-next-line no-console
          console.log(chalk.bold.red('Error: %s'), formatError(err));
          process.on('exit', () => process.exit(1));
        }
      } else {
        // eslint-disable-next-line no-console
        console.log(chalk.bold.red('%s'), formatError(err));
        process.on('exit', () => process.exit(1));
      }
    });
}
