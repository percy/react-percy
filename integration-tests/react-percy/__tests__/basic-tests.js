// eslint-disable-next-line import/no-extraneous-dependencies
import * as percy from 'percy-client';
import { run } from '@percy-io/react-percy/lib/cli';

jest.mock('percy-client');

process.exit = jest.fn();

beforeAll(async () => {
  await run(['--config', 'webpack/webpack.config.js']);
});

it('uploads CSS files', () => {
  expect(percy.uploadResource.mock.calls).toMatchSnapshot();
});

it('creates snapshots for each test case', () => {
  expect(percy.createSnapshot.mock.calls).toMatchSnapshot();
});
