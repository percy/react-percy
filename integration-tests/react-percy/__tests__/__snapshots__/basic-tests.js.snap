// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates snapshots for each test case 1`] = `
Array [
  Array [
    "test-build",
    Array [
      Object {
        "content": "
    <!DOCTYPE html>
    <html lang=\\"en\\">
    <head>
        <meta charset=\\"utf-8\\">
        <link rel=\\"stylesheet\\" href=\\"/styles.css\\" />
    </head>
    <body><div>This is some text</div></body>
    </html>
    ",
        "id": "sha-/example-basic-components-work.html-sha",
        "isRoot": true,
        "mimetype": "text/html",
        "resourceUrl": "/example-basic-components-work.html",
        "sha": "sha-/example-basic-components-work.html-sha",
      },
    ],
    Object {
      "name": "Example - basic components work",
      "widths": Array [],
    },
  ],
  Array [
    "test-build",
    Array [
      Object {
        "content": "
    <!DOCTYPE html>
    <html lang=\\"en\\">
    <head>
        <meta charset=\\"utf-8\\">
        <link rel=\\"stylesheet\\" href=\\"/styles.css\\" />
    </head>
    <body><div>This is some text</div></body>
    </html>
    ",
        "id": "sha-/example-components-with-custom-dimensions-work.html-sha",
        "isRoot": true,
        "mimetype": "text/html",
        "resourceUrl": "/example-components-with-custom-dimensions-work.html",
        "sha": "sha-/example-components-with-custom-dimensions-work.html-sha",
      },
    ],
    Object {
      "name": "Example - components with custom dimensions work",
      "widths": Array [
        320,
        768,
      ],
    },
  ],
]
`;

exports[`uploads CSS files 1`] = `
Array [
  Array [
    "test-build",
    ".text {
    color: red;
}
",
  ],
]
`;
